File: 250708/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.js
const fs = require("fs");
const input = fs.readFileSync(0).toString().trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);
const snakeGrid = input.slice(1, n + 1).map(row => row.split(' ').map(Number));

// Please Write your code here.
const visitedGrid = Array.from({ length: n }, () => Array(m).fill(0));

function dfs(r, c) {
    const [dr, dc] = [[1, 0], [0, 1]];

    for (let i = 0; i < 2; i++) {
        const [nr, nc] = [r + dr[i], c + dc[i]];

        if (canGo(nr, nc)) {
            visitedGrid[nr][nc] = 1;
            dfs(nr, nc);
        }
    }
}

function canGo(r, c) {
    if (!inRange(r, c)) {
        return false;
    }
    if (visitedGrid[r][c]) {
        return false;
    }
    if (!snakeGrid[r][c]) {
        return false;
    }
    return true;
}

function inRange(r, c) {
    return r >= 0 && r < n && c >= 0 && c < m;
}

function canEscape(r, c) {
    dfs(0, 0);

    if (visitedGrid[r - 1][c - 1] === 1) {
        return 1;
    }
    return 0;
}

console.log(canEscape(n, m));

File: 250719/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.py
n, m = map(int, input().split())
snake_grid = [list(map(int, input().split())) for _ in range(n)]

# Please write your code here.
grid = [
    [0] * m
    for _ in range(n)
]

visited = [
    [False] * m
    for _ in range(n)
]

drs, dcs = [1, 0], [0, 1]


def dfs(r, c, depth):
    for dr, dc in zip(drs, dcs):
        nr, nc = r + dr, c + dc

        if can_go(nr, nc):
            visited[nr][nc] = 1
            grid[nr][nc] = depth
            dfs(nr, nc, depth + 1)


def can_go(r, c):
    return in_range(r, c) and not visited[r][c] and snake_grid[r][c]


def in_range(r, c):
    return r >= 0 and r < n and c >= 0 and c < m


dfs(0, 0, 1)

print(1 if grid[n - 1][m - 1] else 0)